cpp/run3ScoutingLooper.C

```diff
@@ -0,0 +1,587 @@
+  std::vector<std::size_t> p(vec.size());
+  std::iota(p.begin(), p.end(), 0);
+  std::sort(p.begin(), p.end(), [&](std::size_t i, std::size_t j){ return compare(vec[i], vec[j]); });
+  return p;
+}
+
+template <typename T>
+void apply_permutation_in_place(std::vector<T>& vec, const std::vector<std::size_t>& p) {
+  std::vector<bool> done(vec.size());
+  for (std::size_t i=0; i<vec.size(); ++i) {
+    if (done[i])
+      continue;
+    done[i] = true;
+    std::size_t prev_j = i;
+    std::size_t j = p[i];
+    while (i != j) {
+      std::swap(vec[prev_j], vec[j]);
+      done[j] = true;
+      prev_j = j;
+      j = p[j];
+    }
+  }
+}
+
+
+float getMiniIsoDR(TLorentzVector muVec, float minDR=0.05, float maxDR=0.2,float kt=10.0) {
+  return std::min(maxDR, std::max(minDR, kt/(float)muVec.Pt()));
+}
+
+std::tuple<float,float,float,float,float> getPFIsolation(TLorentzVector muVec, std::vector<Run3ScoutingParticle> pfs, float maxDR=0.3, float vetoDR=0.01, float maxdz=0.1, bool doMini=false, bool doMindrMuPF=true) {
+  float maxdr, mindrMuPF=1e6;
+  float chIso=0, nhIso=0, phIso=0, puIso=0;
+  if (doMini)
+    maxdr = getMiniIsoDR(muVec);
+  for (auto pf : pfs) {
+    TLorentzVector pfvec;
+    pfvec.SetPtEtaPhiM(pf.pt(),pf.eta(),pf.phi(),0.0);
+    float dr = muVec.DeltaR(pfvec);
+    if (dr<mindrMuPF)
+      mindrMuPF = dr;
+    if (dr<vetoDR || dr>maxDR)
+      continue;
+    bool fromPV = !(abs(pf.vertex())>0 || pf.dz()>maxdz);
+    bool isCh=false, isNh=false, isPh=false, isPU=false;
+    if (fabs(pf.pdgId())==211) {
+      if (fromPV) {
+        isCh = true;
+        chIso = chIso + pf.pt();
+      }
+      else {
+        isPU = true;
+        puIso = puIso + pf.pt();
+      }
+    }
+    if (fabs(pf.pdgId())==130) {
+      isNh  = true;
+      nhIso = nhIso + pf.pt();
+    }
+    if (fabs(pf.pdgId())==22) {
+      isPh  = true;
+      phIso = phIso + pf.pt();
+    }
+  }
+  if (doMindrMuPF)
+    return {chIso,nhIso,phIso,puIso,mindrMuPF};
+  else
+    return {chIso,nhIso,phIso,puIso,1e6};
+}
+
+
+struct SV {
+  std::vector<unsigned int> index, ndof;
+  std::vector<float> x, y, z;
+  std::vector<float> xe, ye, ze;
+  std::vector<float> chi2, prob, chi2Ndof;
+  std::vector<float> lxy, l3d;
+  std::vector<float> mindx, mindy, mindz, mindxy, mind3d;
+  std::vector<float> maxdx, maxdy, maxdz, maxdxy, maxd3d;
+  std::vector<bool> isValid;
+
+  void clear() {
+    index.clear(); ndof.clear();
+    x.clear(); y.clear(); z.clear();
+    xe.clear(); ye.clear(); ze.clear();
+    chi2.clear(); prob.clear(); chi2Ndof.clear();
+    lxy.clear(); l3d.clear();
+    mindx.clear(); mindy.clear(); mindz.clear(); mindxy.clear(); mind3d.clear();
+    maxdx.clear(); maxdy.clear(); maxdz.clear(); maxdxy.clear(); maxd3d.clear();
+    isValid.clear();
+  }
+
+  void sort() {
+    auto comp = sort_permutation(prob, [](float const& a, float const& b){ return a < b; }); // Define permutations based on the prob vector
```

Comment:
IIUC, this is sorting vertices by chi2 probability, putting lowest probability first.
We should possibly revert this, so that the vertices with the highest probability are ranked first?